### ISSUE TITLE
IDLC can't find nested IDLs when using include search list

### ISSUE THREAD

**Issue by sumanth-nirmal** [eclipse-cyclonedds/cyclonedds#844]
## Required information

**Operating system:**
Ubuntu 18.04 LTS

**Version:**
cyclonedds => master (e34d38af35713297f0865a32f8889dca925e2d8c)

**Observed result or behaviour:**
When using nested IDL in an IDL file and using `-I` option of the IDL compiler (`idlc`) it actually fails to find the nested IDL file

**Expected result or behaviour:**
The `-I` options should be correctly honored

**Conditions where it occurred / Performed steps:**

IDL file includes another IDL `Time_.idl` as below which is in `/home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share`

```hpp
#include "builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl"

module foo {
...
```

So invoking the `idlc` as below gives an error 
```sh
/home/sumanth.nirmal/cyclone_iox_ws/cyclonedds/install/bin/idlc -I /home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share -l /home/sumanth.nirmal/cyclone_iox_ws/cyclonedds-cxx/install/lib/libidlcxx.so /home/sumanth.nirmal/gc/apex_ws/build/rosgraph_msgs/rosidl_generator_dds_idl/rosgraph_msgs/msg/dds_cyclone_typefiles/Clock_.idl

Can't open include file "builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl"
    #include "builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl"
1 error in preprocessor.
```

Looking at the preprocessor code, it looks like the path is concatenated instead of copying [here](https://github.com/eclipse-cyclonedds/cyclonedds/blob/e34d38af35713297f0865a32f8889dca925e2d8c/src/tools/idlpp/src/system.c#L2592)

**Comment by k0ekk0ek** [comment]
Hi @sumanth-nirmal! Your suggestion doesn't seem weird, but I can't seem to reproduce the problem myself. I used more-or-less the same paths as you. From the build directory.

`share/builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl`:
```
struct s2 {
  long l2;
};
```

`Clock_.idl`:
```
#include "builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl"

struct s1 {
  long l1;
};
```

And then running `bin/idlc -I share Clock_.idl` works fine. As does `bin/idlc -I $(pwd)/share Clock_.idl`. So while your fix makes sense, I'm not sure it's completely right. It depends on the value of the symbolic link I guess?

**Comment by k0ekk0ek** [comment]
The above comment was meant for #845 of course, but I cannot reproduce whether I use a relative or an absolute symlink, so I must be overlooking something?

**Comment by k0ekk0ek** [comment]
Looking at this again, it seems the real fix will be a little more involved. The original code did this:
```
    } else if (fname) {                             /* Regular file */
        len = strlen( slbuf1);
        strcat( slbuf1, fname);
        deref_syml( slbuf1, slbuf2, slbuf1 + len);
                                /* Symbolic link check of directory */
        if ((len = readlink( slbuf1, slbuf2, PATHMAX)) > 0) {
            /* Dereference symbolic linked file (not directory) */
            *(slbuf2 + len) = EOS;
            cp1 = slbuf1;
            if (slbuf2[ 0] != PATH_DELIM) {     /* Relative path    */
                cp2 = strrchr( slbuf1, PATH_DELIM);
                if (cp2)        /* Append to the source directory   */
                    cp1 = cp2 + 1;
            }
            strcpy( cp1, slbuf2);
        }
    }
```
So, the cat is right if the filename is relative and the copy is right if the filename is absolute. I'll have a look later today.

**Comment by sumanth-nirmal** [comment]
@k0ekk0ek Sorry, I still owe you a true reproducer, but looks like you already figured out the root cause. 

I was using absolute paths, and when debugging I see that the path is appended as below:

- filename (`fname`)=> `builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl`
- directory (`dir`)=> `/home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share`
- slbuf1 =>  `/home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share/builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl`
- slbuf2 =>  `/home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share/builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl`

After `strlcat`, `slbuf1` is ` /home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share/builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl /home/sumanth.nirmal/gc/apex_ws/install/builtin_interfaces/share/builtin_interfaces/msg/dds_cyclone_typefiles/Time_.idl`

This was the reasoning for `strlcpy` in my proposed fix, but as you mentioned we should probably also tackle the scenario with the relative path

**Comment by k0ekk0ek** [comment]
No worries. Can you give https://github.com/k0ekk0ek/cyclonedds/tree/fix-844 a try?

**Comment by sumanth-nirmal** [comment]
@k0ekk0ek #854 works as expected for my scenario

**Comment by sumanth-nirmal** [comment]
Closed through #854
